shopinvader_address/models/res_partner.py

```diff
@@ -0,0 +1,160 @@
+                    ("state", "in", ("done", "sale")),
+                ],
+                limit=1,
+            )
+        )
+        if len(move_id) > 0:
+            raise UserError(
+                _(
+                    "Can not update billing addresses(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
+            )
+
+    def _ensure_shopinvader_shipping_address_not_used(self) -> None:
+        """
+        Check if Shipping Address is used on confirmed sale order
+        """
+        self.ensure_one()
+        move_id = (
+            self.env["sale.order"]
+            .sudo()
+            .search(
+                [
+                    ("partner_shipping_id", "=", self.id),
+                    ("state", "in", ("done", "sale")),
+                ],
+                limit=1,
+            )
+        )
+        if len(move_id) > 0:
+            raise UserError(
+                _(
+                    "Can not delete Shipping address(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
+            )
+
+    # --- Billing ---
+    # Billing addresses is unique and corresponds to authenticated_partner
+
+    def _get_shopinvader_billing_addresses(self) -> "ResPartner":
+        self.ensure_one()
+        return self
+
+    def _create_shopinvader_billing_address(self, vals: dict) -> "ResPartner":
+        raise UserError(_("Creation of billing addresses is not supported"))
+
+    def _update_shopinvader_billing_address(
+        self, vals: dict, address_id: int
+    ) -> "ResPartner":
+        self.ensure_one()
+
+        if address_id != self.id:
+            raise UserError(
+                _(
+                    "Can not update billing addresses(%(address_id)d)"
+                    "because it doesn't correspond to authenticated partner's billing address",
+                    address_id=self.id,
+                )
+            )
+
+        # if billing addresses is already used, it is not possible to modify it
+        # an error will be raised
+        self._ensure_shopinvader_billing_address_not_used()
+
+        self.write(vals)
+
+        return self
+
+    # ---Shipping ---
+
+    def _get_shopinvader_shipping_addresses(
+        self, address_id: int = None
+    ) -> "ResPartner":
+        self.ensure_one()
+        domain = [("type", "=", "delivery"), ("parent_id", "=", self.id)]
+
+        if address_id is not None:
+            domain.append(("id", "=", address_id))
+
+        return self.env["res.partner"].search(domain)
+
+    def _create_shopinvader_shipping_address(self, vals: dict) -> "ResPartner":
+        self.ensure_one()
+        vals.update(
+            {
+                "parent_id": self.id,
+                "type": "delivery",
+            }
+        )
```

Comment:
A good habit to take is to never update these `vals` dictionaries.
Instead take a copy: `vals = dict(vals, parent_id=self.id, type="delivery")`.
In this particular case it was harmless but if you do important changes and the transaction is retried, some Odoo version retry with the modified dictionary.